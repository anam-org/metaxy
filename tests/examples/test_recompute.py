"""Tests for example scripts."""

import os
import subprocess
import sys
from pathlib import Path

import pytest


def test_pipeline(tmp_path):
    """Test that the recompute example runs successfully."""
    # Set up environment with test-specific database path
    env = os.environ.copy()
    test_db = tmp_path / "example_recompute.db"
    env["METAXY_STORES__DEV__CONFIG__DATABASE"] = str(test_db)

    example_dir = Path("examples/src/examples/recompute")

    # Step 1: Setup upstream data
    result = subprocess.run(
        [sys.executable, "-m", "examples.recompute.setup_data"],
        capture_output=True,
        text=True,
        timeout=10,
        env=env,
        cwd=example_dir,
    )
    assert result.returncode == 0, f"Setup failed: {result.stderr}"
    print(result.stdout)

    # Step 2: Run pipeline with STAGE=1
    env["STAGE"] = "1"
    result = subprocess.run(
        [sys.executable, "-m", "examples.recompute.pipeline"],
        capture_output=True,
        text=True,
        timeout=30,
        env=env,
        cwd=example_dir,
    )
    assert result.returncode == 0, (
        f"Stage 1 failed: {result.stderr}\nstdout: {result.stdout}"
    )
    print(result.stdout)
    assert "Pipeline STAGE=1" in result.stdout
    assert "✅ Stage 1 pipeline complete!" in result.stdout

    # Step 3: Run pipeline with STAGE=1 again (demonstrate idempotence)
    result = subprocess.run(
        [sys.executable, "-m", "examples.recompute.pipeline"],
        capture_output=True,
        text=True,
        timeout=30,
        env=env,
        cwd=example_dir,
    )
    assert result.returncode == 0, (
        f"Stage 1 (rerun) failed: {result.stderr}\nstdout: {result.stdout}"
    )
    print("\n--- Stage 1 rerun (should be idempotent) ---")
    print(result.stdout)
    assert "Pipeline STAGE=1" in result.stdout
    # Should not recompute anything
    assert "No changes detected (idempotent)" in result.stdout

    # Step 4: Run pipeline with STAGE=2 (with updated algorithm)
    env["STAGE"] = "2"
    result = subprocess.run(
        [sys.executable, "-m", "examples.recompute.pipeline"],
        capture_output=True,
        text=True,
        timeout=30,
        env=env,
        cwd=example_dir,
    )
    assert result.returncode == 0, (
        f"Stage 2 failed: {result.stderr}\nstdout: {result.stdout}"
    )
    print(result.stdout)
    assert "Pipeline STAGE=2" in result.stdout
    assert "✅ Stage 2 pipeline complete!" in result.stdout
    # When feature_version changes, samples are "new" for that version (Materialized), not "changed" (Recomputed)
    assert (
        "Materialized" in result.stdout
        or "Recomputed" in result.stdout
        or "new samples" in result.stdout
        or "changed samples" in result.stdout
    ), "Child feature should recompute when parent algorithm changes"


@pytest.mark.parametrize("stage", ["1", "2"])
def test_list_features(snapshot, stage):
    """Snapshot test for 'metaxy list features' output at different stages.

    This captures the feature specifications at STAGE=1 and STAGE=2,
    demonstrating how feature versions change when code_version changes.
    """
    # Set up environment
    env = os.environ.copy()
    env["STAGE"] = stage

    # Run metaxy list features command
    result = subprocess.run(
        [sys.executable, "-m", "metaxy.cli.app", "list", "features"],
        capture_output=True,
        text=True,
        timeout=10,
        env=env,
        cwd=Path("examples/src/examples/recompute"),
    )

    assert result.returncode == 0, f"Command failed: {result.stderr}"

    # Snapshot the output
    assert result.stdout == snapshot
