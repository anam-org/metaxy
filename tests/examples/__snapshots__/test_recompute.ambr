# serializer version: 1
# name: test_list_features[1]
  '''
  ---
  examples/parent (version
  0aad9b8a2ea055cde3c4149fc4cde576e6478a982cea75b45c3cd012db43a5e8)
    Fields:
      embeddings (code_version 1, version
  05e66510da58ef37168095b60188849cd6b1f0a4b539d0ac29ffd1e15b756459)
  ---
  examples/child (version
  440ffb028aaa5cb21b155c4ef21debd81f283f99aa91ef58cbe541d71164b44f)
    Feature Dependencies:
      feature=examples/parent columns=None rename=None
    Fields:
      predictions (code_version 1, version
  1905d9e85ceb58b0361e863c7fbfbf5843582f8ed135ea5ce6c24ee2c68c6bb4)
  
  '''
# ---
# name: test_list_features[2]
  '''
  ---
  examples/parent (version
  a007f308d0a852e3fbf80a442bd0089e29eed94efefe85231ef4fc927aa7d737)
    Fields:
      embeddings (code_version 2, version
  3c8d3e9ba031ab3613eb4db0877d3959fce76d94d625335b89bae7fbd4f27add)
  ---
  examples/child (version
  7251e21c32d2d8e35a8ba389a8ce1b597663f206dee0ff55e542a3af1f1665cd)
    Feature Dependencies:
      feature=examples/parent columns=None rename=None
    Fields:
      predictions (code_version 1, version
  bcb950543ea50ba3e19aef4846cbc3d36725e2548040c4327219eb2e75e6d997)
  
  '''
# ---
